File: Leetcode/dfs_recursion_backtracking/784_LetterCasePermutation/Solution.java
class Solution {
    public List<String> letterCasePermutation(String S) {
        List<String> res = new LinkedList<String>();
        char[] chars = S.toCharArray();
        dfs(res, chars, 0);
        return res;
    }
        private void dfs(List<String> res, char[] chars, int index){
            if(index==chars.length){
                res.add(new String(chars));
                return;
            }
            if (Character.isLetter(chars[index])){
                chars[index]^=32;
                dfs(res, chars, index+1);
                chars[index]^=32;
                dfs(res, chars, index+1);
            } else{
                dfs(res, chars, index+1);
            }
        }
}

File: Leetcode/dfs_recursion_backtracking/784_LetterCasePermutation/Solution.py
class Solution(object):
    def letterCasePermutation(self, S):
        """
        :type S: str
        :rtype: List[str]
        """
        res = []
        def dfs(i, S, temp):
            if i==len(S):
                res.append("".join(temp))
                return
            if S[i].isalpha():
                temp.append(S[i].lower())
                dfs(i+1, S, temp)
                temp.pop()
                temp.append(S[i].upper())
                dfs(i+1, S, temp)
                temp.pop()
            else:
                temp.append(S[i])
                dfs(i+1, S, temp)
                temp.pop()
        dfs(0, S, [])
        return res
